File: alg/lru-cache.cs
/*
  Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.
  
  Implement the LRUCache class:
  
  LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
  int get(int key) Return the value of the key if the key exists, otherwise return -1.
  void put(int key, int value) Update the value of the key if the key exists. Otherwise, add the key-value pair to the cache. If the number of keys exceeds the capacity from this operation, evict the least recently used key.
  The functions get and put must each run in O(1) average time complexity.
*/
using System.Runtime.CompilerServices;

public sealed class Node {
    public int key, value;
    public Node next, prev;

    public Node(int key = 0, int value = 0) {
        this.key = key;
        this.value = value;
    }
}

public sealed class LRUCache {
    private readonly Dictionary<int, Node> cache;
    private readonly int capacity;

    private Node head, tail;

    public LRUCache(int capacity) {
        cache = new(capacity);
        this.capacity = capacity;

        head = new();
        tail = new();

        head.next = tail;
        tail.prev = head;
    }
    
    public int Get(int key) {
        if (cache.ContainsKey(key)) {
            Node node = cache[key];
            MoveToHead(node);
            return node.value;
        }
        return -1;
    }
    
    public void Put(int key, int value) {
        if (cache.ContainsKey(key)) {
            Node node = cache[key];
            node.value = value;
            MoveToHead(node);
        } else {
            Node node = new(key, value);
            cache.Add(key, node);
            Add(node);

            if (cache.Count > capacity) {
                Node lru = PopTail();
                cache.Remove(lru.key);
            }
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void MoveToHead(Node node) {
        Remove(node);
        Add(node);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void Remove(Node node) {
        Node next = node.next;
        Node prev = node.prev;

        prev.next = next;
        next.prev = prev;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void Add(Node node) {
        node.prev = head;
        node.next = head.next;

        head.next.prev = node;
        head.next = node;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private Node PopTail() {
        Node rem = tail.prev;
        Remove(rem);
        return rem;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */

File: alg/lru-cache.rs
use std::collections::HashMap;
use std::cell::RefCell;
use std::rc::{Rc, Weak};

struct LRUCache {
    capacity: usize,
    cache: HashMap<i32, Rc<RefCell<CacheNode>>>,
    head: Rc<RefCell<CacheNode>>,
    tail: Rc<RefCell<CacheNode>>,
}

struct CacheNode {
    key: i32,
    value: i32,
    prev: Option<Weak<RefCell<CacheNode>>>,
    next: Option<Rc<RefCell<CacheNode>>>,
}

impl CacheNode {
    fn new(key: i32, value: i32) -> Self {
        Self { key, value, prev: None, next: None }
    }
}

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl LRUCache {

    fn new(capacity: i32) -> Self {
        let capacity = capacity as usize;
        let head = Rc::new(RefCell::new(CacheNode::new(0, 0)));
        let tail = Rc::new(RefCell::new(CacheNode::new(0, 0)));
        head.borrow_mut().next = Some(tail.clone());
        tail.borrow_mut().prev = Some(Rc::downgrade(&head));

        Self {
            capacity,
            cache: HashMap::with_capacity(capacity),
            head,
            tail,
        }
    }
    
    fn get(&mut self, key: i32) -> i32 {
        if let Some(node) = self.cache.get(&key).cloned() {
            self.move_to_head(node.clone());
            return node.borrow().value;
        }
        -1
    }
    
    fn put(&mut self, key: i32, value: i32) {
        if let Some(node) = self.cache.get(&key).cloned() {
            node.borrow_mut().value = value;
            self.move_to_head(node);
        } else {
            let node = Rc::new(RefCell::new(CacheNode::new(key, value)));
            self.cache.insert(key, node.clone());
            self.add(node.clone());
            
            if self.cache.len() > self.capacity {
                if let Some(removed) = self.pop_tail() {
                    self.cache.remove(&removed.borrow().key);
                }
            }
        }
    }

    fn move_to_head(&self, node: Rc<RefCell<CacheNode>>) {
        self.remove(node.clone());
        self.add(node);
    }

    fn add(&self, node: Rc<RefCell<CacheNode>>) {
        let mut head_next = self.head.borrow_mut().next.take().unwrap();
        node.borrow_mut().prev = Some(Rc::downgrade(&self.head));
        node.borrow_mut().next = Some(head_next.clone());
        head_next.borrow_mut().prev = Some(Rc::downgrade(&node));
        self.head.borrow_mut().next = Some(node);
    }

    fn remove(&self, node: Rc<RefCell<CacheNode>>) {
        let prev = node.borrow_mut().prev.take().unwrap().upgrade().unwrap();
        let next = node.borrow_mut().next.take();
        if let Some(next) = next {
            next.borrow_mut().prev = Some(Rc::downgrade(&prev));
            prev.borrow_mut().next = Some(next);
        } else {
            prev.borrow_mut().next = None;
        }
    }

    fn pop_tail(&self) -> Option<Rc<RefCell<CacheNode>>> {
        let tail_prev = self.tail.borrow_mut().prev.take()?.upgrade()?;
        self.remove(tail_prev.clone());
        Some(tail_prev)
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache::new(capacity);
 * let ret_1: i32 = obj.get(key);
 * obj.put(key, value);
 */
